### ISSUE TITLE
Deployed Web App Not Secure with Valid Certificate

### ISSUE THREAD

**Issue by shash222** [envoyproxy/envoy#19552]
I have a web app written in the following tech stack:

- Web Server: Nginx
- UI: Flutter/Dart
- Backend: Go
- Communication Protocol: gRPC/gRPC-Web
- HTTP1.1 to HTTP2 Proxy: Envoy Proxy (Dockerized)
- SSL Certificate Provider: LetsEncrypt Certbot
- Environment: Ubuntu 20.10

When I access the site, initially I see the lock on chrome showing the site is secure, however when the page actually loads, I see `Not Secure`, though the certificate is valid according to chrome and I can't figure out why.

As mentioned above, I am using a certificate generated by LetsEncrypt's Certbot which I have used a bunch of times in the past, though with REST instead of gRPC (Nginx web server, node backend), and have never had this issue, so I'm assuming it's an issue with the way I have configured Envoy or my backend service with TLS.

Here is my `Dockerfile`:

    FROM envoyproxy/envoy:v1.20-latest
    COPY ./envoy.yaml /etc/envoy/envoy.yaml
    ADD ./ssl_certs/privkey.pem /etc/envoy/privkey.pem
    ADD ./ssl_certs/fullchain.pem /etc/envoy/fullchain.pem
    CMD /usr/local/bin/envoy -c /etc/envoy/envoy.yaml -l debug

My `envoy.yaml`:

    static_resources:
      listeners:
        - name: listener_0
          address:
            socket_address: { address: 0.0.0.0, port_value: 9000 }
          filter_chains:
          - filters:
            - name: envoy.http_connection_manager
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
                codec_type: auto
                stat_prefix: ingress_http
                route_config:
                  name: local_route
                  virtual_hosts:
                  - name: local_service
                    domains: ["*"]
                    routes:
                    - match: { prefix: "/" }
                      route:
                        cluster: greeter_service
                        max_grpc_timeout: 0s
                    cors:
                      allow_origin_string_match:
                      - prefix: "*"
                      allow_methods: GET, PUT, DELETE, POST, OPTIONS
                      allow_headers: keep-alive,user-agent,cache-control,content-type,content-transfer-encoding,custom-header-1,x-accept-content
    -transfer-encoding,x-accept-response-streaming,x-user-agent,x-grpc-web,grpc-timeout
                      max_age: "1728000"
                      expose_headers: custom-header-1,grpc-status,grpc-message
                http_filters:
                - name: envoy.filters.http.cors
                - name: envoy.filters.http.grpc_web
                - name: envoy.filters.http.router
            transport_socket:
              name: envoy.transport_sockets.tls
              typed_config:
                "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext
                common_tls_context:
                  tls_certificates:
                  - certificate_chain: {filename: "/etc/envoy/fullchain.pem"}
                    private_key: {filename: "/etc/envoy/privkey.pem"}
      clusters:
        - name: greeter_service
          connect_timeout: 0.25s
          type: logical_dns
          lb_policy: round_robin
          http2_protocol_options: {}
          load_assignment:
            cluster_name: cluster_0
            endpoints:
              - lb_endpoints:
                  - endpoint:
                      address:
                        socket_address:
                          address: 127.0.0.1
                          port_value: 9001
          transport_socket:
            name: envoy.transport_sockets.tls
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
              common_tls_context:
                tls_certificates:
                - certificate_chain: {filename: "/etc/envoy/fullchain.pem"}
                  private_key: {filename: "/etc/envoy/privkey.pem"}

And my Go controller:

    import (
            "fmt"
            "log"
            "net"
            "crypto/tls"
            "google.golang.org/grpc"
            "google.golang.org/grpc/credentials"
            protos "example.com/protos"
            "example.com/repositories"
            "example.com/services"
    )
    const (
            port = 9001
    )
    var (
            grpcServer *grpc.Server
    )
    func loadTLSCredentials() (credentials.TransportCredentials, error) {
            // Load TLS certificate and key
            serverCert, err := tls.LoadX509KeyPair("/etc/letsencrypt/live/example.com/fullchain.pem", "/etc/letsencrypt/live/example.com/privkey.pem")
            if err != nil {
                    return nil, err
            }
            // Create TLS credentials
            config := &tls.Config {
                    Certificates: []tls.Certificate{serverCert},
                    ClientAuth:   tls.NoClientCert,
            }
            return credentials.NewTLS(config), nil
    }
    // StartServer starts gRPC server
    func StartServer() {
            log.Println("Starting server")
            tlsCredentials, err := loadTLSCredentials()
            if err != nil {
                    log.Fatalf("Unable to load TLS Credentials: %v\n ", err);
            }
            listener, err := net.Listen("tcp4", fmt.Sprintf(":%v", port))
            if err != nil {
                    log.Fatalf("Unable to listen to port %v\n%v\n", port, err)
            }
            repositories.ConnectToMongoDB()
            grpcServer = grpc.NewServer(
                    grpc.Creds(tlsCredentials),
                    // grpc.UnaryInterceptor(interceptor.Unary()),
                    // grpc.StreamInterceptor(interceptor.Stream()),
            )
            registerServices()
            if err = grpcServer.Serve(listener); err != nil {
                    log.Fatalf("Failed to serve gRPC\n%v\n", err)
            }
    }

All of my services are using the same `fullchain.pem` and `privkey.pem`, I had copied it into `ssl_certs` manually which is why I can access it in 2 places.

What am I doing wrong? At first, I was unable to call my backend service when Nginx was set to listen to HTTPS only and I had not yet configured Envoy or my backend service with HTTPS, which is what I had expected since the webapp itself had SSL enabled. Once I added SSL to Envoy and my backend service, as shown above, the API calls were successful and I see the correct responses, however the site is marked `Not secure`, even with a valid certificate. I am at a loss. All of the API calls being made are also `https` calls, and not `http`.

This is my first time posting an issue, I apologize if this is not the correct way or place for this. Please let me know if that's the case and I can update/move it accordingly.

**Comment by shash222** [comment]
I was hoping to follow up on this, I haven't been able to find a solution yet. Any ideas why this is happening?

**Comment by ggreenway** [comment]
Trying to make sure I understand correctly: everything works EXCEPT chrome says it is "not secure"?

**Comment by shash222** [comment]
@ggreenway Correct. After attaching the certbot-generated SSL/TLS certificate, I am able to load the web app on `https`, and am able to make gRPC calls as well over https, and I get the expected responses. The only issue is that chrome says "Not secure"

**Comment by ggreenway** [comment]
I don't know what criteria chrome uses to determine whether a site is "secure" or not. I wonder if it's related to Certificate Transparency?

**Comment by shash222** [comment]
So I tried loading the application on Chrome, Chrome Incognito, Firefox, and Edge and got different behaviors for each one:

Chrome:
As mentioned above, page loads, API calls are successful, but Chrome says "Not Secure"

Chrome Incognito:
Same as Chrome

Firefox:
Application loads and site is secure, but gRPC calls failed
`OPTIONS` call failed with `CORS Failed`
`POST` call failed

Microsoft Edge
Application loads and site is secure, but gRPC calls failed
Both `OPTIONS` and `POST` call failed with `ERR_CERT_COMMON_NAME_INVALID`



I checked the warning on the Firefox page, and it looked like that had failed the API calls because I was calling the API on the server IP instead of the domain. This was an issue because the certificate was only valid for the domain.

I updated the client to use the domain instead of the server IP, and it looks like the browsers are now happy with the change. Hopefully this helps anyone else that's facing the same issue.


TLDR
The issue was not a server-side issue, it looks like my configuration was correct for both envoy and my go server. The issue lied with the URL I was calling through gRPC. I was calling my server's IP address instead of my domain, which is the only address that my certificate was configured for, but updating that fixed the issue.
